feat: add runtime and runtime-macros libraries

Cai1Hsu opened the pull request:
Unfinished. This PR is intended to replace the existing HAL.

- Create `runtime-macros` library for proc-macro implementation
- Create `runtime` library to provide runtime support, mainly boot and interrupt support
- Add `rust_main` proc-macro for creating a main function compatible with both baremetal and std environments.

std test passing, can be tested with:

<details>
<summary>
code
</summary>

- Cargo.toml
```toml
[package]
name = "program"
version = "0.1.0"
edition = "2024"

[dependencies]
runtime = { path = "bakaos/libraries/runtime" }
```

- src/main.rs
```rust
#[runtime::rust_main]
fn main() {
    println!("Hello, world");
}
```
</details>

TODO:
- [ ] add `runtime-int` for interrupt and privilege switch support (Not in this PR)

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

- New Features
  - Introduced modular runtime and standard-compat libraries with no_std support.
  - Added RISC-V64 bare-metal boot/entry pipeline, CPU-local storage, and BSS clearing.
  - Provided serial output macros for formatted printing.
  - Implemented a default panic handler with console output.
  - Added an attribute-based main wrapper for a unified entry across environments.
  - Exposed system shutdown for bare-metal targets.

- Chores
  - Expanded workspace to include new runtime, std-compat, and macro crates.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Cai1Hsu committed:
feat(runtime): add runtime and runtime-macros libraries

- Create runtime-macros library for proc-macro implementation
- Create runtime library to provide runtime support
- Add rust_main proc-macro for creating a main function compatible with both baremetal and std environments

Cai1Hsu commented on libraries/runtime-macros/src/lib.rs:
```diff
@@ -0,0 +1,101 @@
+            // The runtime crate is expected to provide this function:
+            // We call it and pass the user's main implementation.
+            runtime::rust_load_main(#impl_ident);
+        }
```
the `deprecated` attribute is to warn developers not to use main directly

Cai1Hsu replied:
yeah, i'll fix them later

Cai1Hsu replied:
the original deprecated syntax works, and `note =` requires a string literal, and throws a compilation error, i'll keep it for now.

Cai1Hsu commented on libraries/runtime-macros/src/lib.rs:
```diff
@@ -0,0 +1,101 @@
+        }
+
+        #[doc(hidden)]
+        mod __runtime_generated {
+            /// generated entry for runtime crate to load the user's main
+            #[unsafe(no_mangle)]
+            extern "Rust" fn rust_main_entry() {
+                #[allow(deprecated)]
+                let _ = super::main();
+            }
+        }
```
Rust edition guide says:

```
Rust 1.82 added the ability in all editions to mark certain attributes as unsafe to indicate that they have soundness requirements that must be upheld.[1](https://doc.rust-lang.org/nightly/edition-guide/rust-2024/unsafe-attributes.html#footnote-RFC3325) The syntax for an unsafe attribute looks like this:

// SAFETY: there is no other global function of this name
#[unsafe(no_mangle)]
pub fn example() {}
Marking the attribute with unsafe highlights that there are safety requirements that must be upheld that the compiler cannot verify on its own.

Starting with the 2024 Edition, it is now required to mark these attributes as unsafe. The following section describes the safety requirements for these attributes.
```

`unsafe` is required.

Cai1Hsu replied:
`unsafe` works in the current version, and updating to the 2024 edition is in the plan

Cai1Hsu commented on libraries/runtime/src/lib.rs:
```diff
@@ -0,0 +1,13 @@
+    // TODO
+
+    main()
+}
```
https://github.com/caiyih/bakaos/pull/107#discussion_r2343469870

unsafe is required for no_mangle since Rust 2024

Cai1Hsu committed:
refactor(runtime): rename std module to hosted

- Rename 'std' module to 'hosted' to better reflect its purpose
- Update import in lib.rs to use new module name

Cai1Hsu committed:
style(runtime-macros): remove unnecessary semicolon

- Remove redundant semicolon in the runtime-macros expansion

Cai1Hsu committed:
refactor(runtime-macros): update deprecated attribute in generated main function

- Change `#[deprecated]` attribute to `#[deprecated(note = ...)]` format
- Improve deprecation warning message for generated entry points

Cai1Hsu committed:
build(runtime-macros): add proc-macro-crate dependency and improve runtime crate resolution

- Add proc-macro-crate dependency to resolve runtime crate path
- Handle dependency renames and fallback to 'runtime' if not found
- Update rust_main macro to use dynamic crate path

Cai1Hsu committed:
refactor(runtime-macros): update deprecated attribute usage in main function

- Replace `#[doc = ""]` with `#[deprecated = ""]` for consistency
- This change improves code readability and maintains consistent formatting
  for deprecated attributes across the project

Cai1Hsu committed:
feat(runtime): make runtime-macros optional and add feature flags

- Make runtime-macros an optional dependency
- Add feature flags to control runtime functionality
- Create a new entry module for the boot feature
- Update lib.rs to conditionally include entry module

Cai1Hsu committed:
build(runtime): update Cargo.toml to make `binary_crate` feature optional

- Remove `binary_crate` from default features
- Add comment explaining the usage of `binary_crate` feature

Cai1Hsu commented on libraries/runtime/src/lib.rs:
```diff
@@ -0,0 +1,8 @@
+mod baremetal;
+mod hosted;
+
+#[cfg(feature = "boot")]
+mod entry;
+
+#[cfg(feature = "boot")]
+pub use entry::*;
```
`boot` module is provided only for binary crates, or those containing a main/entry function. This library not only boots the user's code, but also plays a role like the `std` library.

Cai1Hsu committed:
build(runtime): add documentation for std feature in Cargo.toml

Cai1Hsu committed:
feat(runtime): add no_std support and std compatibility layer

- Add std_compat module to provide a compatibility layer for no_std environments
- Re-export core and alloc items to mimic std functionality
- Update lib.rs to conditionally include std or std_compat based on feature flags
- Add alloc crate as a dependency for no_std environments

Cai1Hsu committed:
refactor(runtime): move std_compat module declaration under conditional compilation

Cai1Hsu commented on libraries/runtime/src/std_compat.rs:
```diff
@@ -0,0 +1,55 @@
+#[allow(deprecated, deprecated_in_future)]
+pub use ::core::u8;
+#[allow(deprecated, deprecated_in_future)]
+pub use ::core::usize;
```
these lines are copied from the std's code directly. I'm trying not to modify them to keep compatibility when we generate it instead of writing it manually

Cai1Hsu committed:
feat(runtime): add sync module to no_std compatibility layer

Cai1Hsu committed:
refactor(std_compat): remove deprecated numeric constants from re-exports

Cai1Hsu committed:
chore(runtime-macros): disable coverage

Cai1Hsu committed:
build(runtime): add hermit-sync dependency and expose it

- Add hermit-sync = "0.1.6" to Cargo.toml dependencies
- Expose hermit-sync in runtime's public API

Cai1Hsu committed:
build(runtime): implement conditional compilation for runtime_std

- Add build.rs script to manage conditional compilation
- Introduce runtime_std cfg flag based on environment variables
- Update lib.rs to use runtime_std instead of feature gates
- Adjust module visibility and usage based on runtime_std

Cai1Hsu committed:
feat(runtime): add RISC-V architecture support with boot and CPU management

Cai1Hsu committed:
fix(runtime): clean up unused imports and correct imports

Cai1Hsu committed:
fix(runtime): update hermit_sync import to avoid wildcard usage

Cai1Hsu commented on libraries/runtime/src/lib.rs:
```diff
@@ -0,0 +1,33 @@
+#[cfg(runtime_std)]
+pub use ::std::*;
+
+#[cfg(feature = "boot")]
+mod entry;
+
+#[cfg(feature = "boot")]
+pub use entry::*;
```
the `runtime::rust_main' function is used for the `main` function, so only the binary crate requires it.

Cai1Hsu commented on libraries/runtime/src/baremetal/cpu/mod.rs:
```diff
@@ -0,0 +1,49 @@
+    let cls = alloc_frame(layout);
+
+    // Copy the template into the newly allocated memory
+    unsafe { cls.copy_from_nonoverlapping(template_start, cls_len) };
```
it has

Cai1Hsu committed:
build(runtime): fix cargo commands in build script

Cai1Hsu committed:
fix(build): replace hyphen with underscore in feature flags

- Update `feature_env` function to replace '-' with '_' in feature flags
- Ensure compatibility with cargo feature names containing hyphens

Cai1Hsu committed:
fix(runtime): add safety documentation to memory initialization function

Cai1Hsu committed:
refactor(runtime): reorganize module structure and improve boot feature handling

Cai1Hsu commented on libraries/runtime/src/baremetal/mod.rs:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    *MEMORY_START.lock() = __ekernel as usize;
+}
```
if static works, will `static __foo: ();` be better?

Cai1Hsu committed:
feat(runtime): introduce symbol_addr macro for safer symbol address retrieval

Cai1Hsu committed:
fix(runtime): simplify feature condition checks in mod.rs

Cai1Hsu commented on libraries/runtime/src/baremetal/cpu/cls.rs:
```diff
@@ -0,0 +1,30 @@
+}
+
+#[link_section = ".cls"]
+pub(crate) static mut CPU0: CpuLocalStorage = CpuLocalStorage {
+    cpu_id: 0,
+    local_base: core::ptr::null_mut(),
+};
```
the `.cls` section serves as a template. CPU's local storage is allocated by the frame allocator, and then copied from the `.cls` region. So this variable won't be actually modified, i'll remove `mut`.

Cai1Hsu committed:
fix(runtime): remove unsafe mut from CPU0 static declaration

Cai1Hsu commented on libraries/runtime/src/baremetal/cpu/mod.rs:
```diff
@@ -0,0 +1,49 @@
+
+pub(crate) fn alloc_cpu_local_storage(cpuid: u32) -> NonNull<CpuLocalStorage> {
+    let template_start = NonNull::new(__scls as *mut u8).unwrap();
+    let template_end = NonNull::new(__ecls as *mut u8).unwrap();
```
I don't think this is an issue, the symbol is provided in the linker script, and should never be null unless the universe ray happens to flip all set bits of the register

Cai1Hsu committed:
fix(runtime): add safety checks and documentation for CPU local storage access

Cai1Hsu committed:
docs(runtime): enhance safety documentation for clear_bss_range function

Cai1Hsu committed:
refactor(runtime): improve shutdown

Cai1Hsu committed:
fix(registers): add safety documentation for get and set register functions

Cai1Hsu committed:
fix(cpu): update to use TP register for CPU local storage management

Cai1Hsu committed:
fix(boot): correct address calculation by changing OR to ADD for high address continuation

Cai1Hsu commented on libraries/runtime/src/baremetal/panic.rs:
```diff
@@ -0,0 +1,22 @@
+use crate::{panic::PanicInfo, println};
+
+#[panic_handler]
+fn __rust_begin_unwind(info: &PanicInfo) -> ! {
+    unsafe extern "Rust" {
+        fn __panic_handler_impl(info: &PanicInfo) -> !;
+    }
+
+    unsafe { __panic_handler_impl(info) }
+}
```
Direct SBI call is heavily coupled with RISC-V code. there's no SMP support currently, so i don't think it's a big deal.

Cai1Hsu committed:
refactor(cpu): ensure `get_cls_ptr` accessible even `boot` or `binary_crate` not enabled

Cai1Hsu committed:
feat(runtime): reexport `utilities` crates as `utils`

caiyih commented on kernel/Cargo.toml:
```diff
@@ -6,6 +6,7 @@ edition = "2021"
 [dependencies]
 log = "0.4"
 hermit-sync = "0.1.6"
+runtime = { path = "dependencies/libraries/runtime" }
```
the line should be here at this time, i'll remove it and add it back when everything is ready

Cai1Hsu committed:
fix(registers): add missing options to inline assembly for safety

caiyih commented on libraries/runtime/src/baremetal/arch/riscv64/registers.rs:
```diff
@@ -0,0 +1,68 @@
+                #[inline(always)]
+                pub unsafe fn [<set_ $name>](val: usize) {
+                    ::core::arch::asm!(concat!("mv ", stringify!($name), ", {}"), in(reg) val);
+                }
```
`volatile` is deprecated and removed in Rust. `nomem` is enough

Cai1Hsu committed:
fix(cpu): update documentation for `get_cls_ptr` to clarify TP register usage

Cai1Hsu commented on libraries/runtime/Cargo.toml:
```diff
@@ -0,0 +1,29 @@
+[features]
+default = []
+
+# Whether the Rust's standard library is available.
+# If enabled, the std crate is used.
+# if disabled, the core crate, alloc and some custom implementations are used to simulate the std crate
+std = []
+
```
`build.rs` checks `std` feature and outputs `runtime_std` cfg

Cai1Hsu committed:
feat(runtime): add serde_json as a build dependency and improve std feature checks

Cai1Hsu committed:
feat(runtime): implement symbol_ptr! macro for accessing global symbols

- Replace symbol_addr! with symbol_ptr! macro
- Support string literals for symbol names, allowing special characters
- Add type casting support for symbols
- Include unit tests for symbol_ptr! functionality

Cai1Hsu committed:
fix(utils): correct macro reference in symbol_ptr! for consistency

Cai1Hsu committed:
refactor(runtime): make symbol_ptr! macro `unsafe` required

- Add safety documentation to the symbol_ptr! macro
- Require unsafe block to call the macro, making symbol access explicit
- Use const fn to avoid runtime overhead while improving safety
- Update usage in cpu local storage allocation to use unsafe block

Cai1Hsu committed:
fix(utils): add null pointer check in symbol_ptr! macro for safety

Cai1Hsu committed:
Merge branch 'master' into generic-runtime

Cai1Hsu committed:
feat(runtime): split runtime-std library for no_std compatibility

Cai1Hsu commented on libraries/runtime-std/build.rs:
```diff
@@ -0,0 +1,52 @@
+    let json: Value = serde_json::from_str(&output).unwrap();
+
+    json["metadata"]["std"].as_bool().unwrap()
+}
```
The build script can probably be removed, [this comment](https://github.com/rust-lang/rust/issues/77730#issuecomment-705885554) mentioned that `target_os = "none"` means baremetal explicitly.

also mentioned here: https://clang.llvm.org/docs/CrossCompilation.html#target-triple.

Cai1Hsu committed:
refactor(runtime): remove build script and update std feature conditions
